analysis/pseudobulk-bulk-prediction/model-notebooks/build-assess-models.Rmd

```diff
@@ -0,0 +1,260 @@
+    fit_name = fit_name,
+    conv_code = ifelse(is.null(code), 0, code),
+    sigma = sigma,
+    aic = AIC(fit),
+    bic = BIC(fit)
+  )
+  
+}
+
+
+# Extract data frame of residuals and fitted values from a given model
+extract_residuals <- function(fit) {
+  data.frame(
+    residuals = unname(resid(fit)), 
+    fitted_values = unname(fitted(fit))
+  )
+}
+```
+
+
+## Perform modeling
+
+We'll now read in all datasets and build all models. 
+
+```{r, warning = FALSE}
+all_models <- purrr::map2(
+  pseudobulk_files, 
+  bulk_counts,
+  \(pseudo_file, bulk_df) {
+    prepare_data(pseudo_file, bulk_df, exclude_samples) |>
+      model_data()
+  })
+
+# clean up
+rm(bulk_counts)
+```
+
+### Fit statistics
+
+Next, we want to look at some of the fit statistics, including whether the random-effects models converged.
+In the `conv_code` column, `-1` indicates the model did not converge and `0` indicates that it did.
+
+```{r}
+all_models |>
+  purrr::map(
+    \(model_list) {
+      purrr::imap(model_list, extract_fit) |>
+        purrr::list_rbind()
+    }
+  ) |>
+  purrr::list_rbind(names_to = "project_id")
+```
+
+* Across all projects, model AIC and BIC each show improvements as model complexity increases
+* $\sigma$ is roughly the same for the plain and additive models, but it is much lower for the interaction models across the board
+  * $\sigma$ shows the largest relative drop from additive to interaction for `SCPCP000017`, which from previous exploration shows the weakest relationship between bulk and pseudobulk (see `../exploratory-notebooks/compare-bulk-pseudobulk.Rmd`)
+* The interaction models do not converge for projects `SCPCP000001`, `SCPCP000002`, and `SCPCP000006`
+  * These projects had the tightest relationships between bulk and pseudobulk in `../exploratory-notebooks/compare-bulk-pseudobulk.Rmd` and similarly the most similar slopes when looking at individual samples.
+  This observation is therefore somewhat consistent with previous observations
+
+### Residual plots
+
+Next we'll look at the residual plots for each model to see if we can detect anything else there.
+We'll use binning in these plots to get a sense of point density as well.
+For this, we collapse all points above the upper scale bound so that we can visualize smaller values in the binning.
+
+```{r}
+all_models <- readr::read_rds(here::here("all_models.rds"))
+
+residuals_df <- all_models |>
+  purrr::map(
+    \(model_list) {
+        purrr::map(model_list, extract_residuals) |>
+          purrr::list_rbind(names_to = "model_name") 
+  }) |>
+  purrr::list_rbind(names_to = "project_id") |>
+  dplyr::mutate(
+    model_name = forcats::fct_relevel(
+      model_name, "fit_plain", "fit_additive", "fit_interaction"
+    )
+  )
+```
+
+
+```{r fig.height=10, fig.width=10}
+ggplot(residuals_df) + 
+  aes(x = fitted_values, y = residuals) + 
+  geom_bin_2d(binwidth = 0.25) + 
+  scale_fill_viridis_c(limits = c(0, 400), oob = scales::squish, option = "inferno") +
```

Comment:
Color palette looks good. I wonder if 400 may be a bit low of a cutoff, but I'm assuming you played around with it, so I'm going to trust your judgement.

Reply:
Can confirm there was much fiddling!